query/PRoST-query/src/main/java/translator/Translator.java

```diff
@@ -174,67 +172,44 @@ public Node buildTree(final List<Triple> triples, final List<Var> projectionVars
 						joinedGroups.remove(object);
 					}
 				}
-
 				for (final Triple triple : joinedGroups.get(biggestGroupKey).getIwptGroup()) {
 					final String subject = triple.getSubject().toString();
-					joinedGroups.get(subject).removeWptTriple(triple);
+					joinedGroups.get(subject).getWptGroup().remove(triple);
 					if (joinedGroups.get(subject).getIwptGroup().size()
 							+ joinedGroups.get(subject).getWptGroup().size() == 0) {
 						joinedGroups.remove(subject);
 					}
 				}
-				// create node
-				if (joinedGroups.get(biggestGroupKey).getIwptGroup().size()
-						+ joinedGroups.get(biggestGroupKey).getWptGroup().size() >= minimumGroupSize) {
-					nodesQueue.add(new JptNode(joinedGroups.get(biggestGroupKey), prefixes));
-				} else {
-					for (final Triple t : joinedGroups.get(biggestGroupKey).getIwptGroup()) {
-						final String tableName = Stats.getInstance().findTableName(t.getPredicate().toString());
-						final Node newNode = new VpNode(new TriplePattern(t, prefixes), tableName);
-						nodesQueue.add(newNode);
-					}
-					for (final Triple t : joinedGroups.get(biggestGroupKey).getWptGroup()) {
-						final String tableName = Stats.getInstance().findTableName(t.getPredicate().toString());
-						final Node newNode = new VpNode(new TriplePattern(t, prefixes), tableName);
-						nodesQueue.add(newNode);
-					}
-				}
+				createJwptVpNodes(joinedGroups.get(biggestGroupKey), nodesQueue);
 
 				joinedGroups.remove(biggestGroupKey);
 			}
 		} else if (usePropertyTable && !useInversePropertyTable) {
-			// RPT disabled
-			final HashMap<String, List<Triple>> subjectGroups = getSubjectGroups(triples);
-
 			logger.info("WPT and VP models only");
-
-			// create and add the proper nodes
-			for (final String subject : subjectGroups.keySet()) {
-				createPtVPNode(subjectGroups.get(subject), nodesQueue);
+			final HashMap<String, List<Triple>> subjectGroups = getSubjectGroups(triples);
+			for (final List<Triple> triplesGroup : subjectGroups.values()) {
+				createPtVPNode(triplesGroup, nodesQueue);
 			}
 		} else if (!usePropertyTable && useInversePropertyTable) {
-			// PT disabled
-			final HashMap<String, List<Triple>> objectGroups = getObjectGroups(triples);
-
 			logger.info("IWPT and VP only");
-			// create and add the proper nodes
-			for (final String object : objectGroups.keySet()) {
-				createRPtVPNode(objectGroups.get(object), nodesQueue);
+			final HashMap<String, List<Triple>> objectGroups = getObjectGroups(triples);
+			for (final List<Triple> triplesGroup : objectGroups.values()) {
+				createRPtVPNode(triplesGroup, nodesQueue);
 			}
 		} else if (usePropertyTable && useInversePropertyTable) {
-			// RPT, PT, and VP enabled
+			logger.info("WPT, IWPT, and VP models only");
+
 			final HashMap<String, List<Triple>> objectGroups = getObjectGroups(triples);
 			final HashMap<String, List<Triple>> subjectGroups = getSubjectGroups(triples);
-			logger.info("WPT, IWPT, and VP models only");
 
```

Comment:
The size of each group will change when we create the nodes for the elements of the other group, We can either have a sorted list or search for the biggest group again. It's actually O(log n), since we are removing items from the lists at every loop. 

A sorted map also has log n complexity for put/remove operations. For this case that we have, both ways will have the same complexity. Keeping a sorted map has an added overhead to maintain it.

Reply:
cf6d3aa3625ffaa33c861da4e7258dc8ae82b29a

Reply:
> 

The size of each group will change when we create the nodes for the elements of the other group, We can either have a sorted list or search for the biggest group again. It's actually O(log n), since we are removing items from the lists at every loop. A sorted map also has log n complexity for put/remove operations. For this case that we have, both ways will have the same complexity. Keeping a sorted map has an added overhead to maintain it.

Can you explain this in more details because I didn't understand how both variants are with the same complexity?

Reply:
HashMap:

creating groups: O(n)

iterate over groups: (n)
	getting largest key: n
	removing triples: n
	removing group: 1
therefor => While loop: O(n^2)

HashTree:

creating groups: O(n log n)

iterate over groups: (n)
	getting largest key:  log n
	removing triples and updating groups: n log n
	removing group: log n
therefore => While loop: O(n^2 n log n) => O(n^2)


Again, n is the number of triple patterns. I don't see the importance of discussing time complexity for such a small n.  This will all be done in ms, only once per query.